### PR TITLE
feat(cds): 预览访问自动唤醒 + 诊断页链接修复

### PR DESCRIPTION
**Pull request by inernoro**
## 摘要

实现「预览访问自动唤醒」：调度器降温的分支（容器被 `docker stop` 保留、未删除）在真实页面导航访问时自动 `docker restart` 唤醒，展示等待页后落到真实应用，避免用户撞上「回控制台手动部署」的诊断死页。同时修复诊断页「返回 CDS 控制台」「查看加载页预览」两个按钮使用相对路径、导致落到预览子域而非控制台域名的问题。

## 改动 diff

**CDS 核心**：
- `cds/src/services/proxy.ts`：新增 `setOnReviveCooled()` 回调注册、`shouldAutoWakeCooled()` 资格判定、`triggerCooledWake()` 去重触发；在 `routeToBranch` 的非运行/非加载分支处插入「仅顶层导航 + 仅调度器降温」才触发唤醒的拦截；新增 `dashboardBaseUrl()` 并把两个诊断页（`serveDeployErrorLightPillarPage` / `serveBranchGoneFallback`）的按钮链接改用 `dashboardDomain || mainDomain` 绝对地址
- `cds/src/index.ts`：实现 `proxyService.setOnReviveCooled()` 回调——同步翻 `restarting` 状态（首个 await 前）让等待页立刻有进度、逐服务 `restartServiceInPlace`、`branchOperationCoordinator` 租约（kind=`restart`）、成功后 `markHot`+刷新 `lastAccessedAt`、活动日志、失败/异常回退；`CDS_PREVIEW_AUTOWAKE=0` kill-switch 整体关闭
- `cds/tests/services/proxy-auto-wake.test.ts`（新增）：7 个 vitest 用例守住 scoping（仅调度器降温 idle 分支 + 顶层导航才唤醒；用户停止/报错/无服务/静态资源不唤醒；并发去重；未接线不报错）

**文档**：
- `doc/debt.cds-nginx-loading-pages.md`：新增 D4 记录自动唤醒的已知边界与完成标准
- `changelogs/2026-06-23_cds-preview-error-page-links.md`（新增）：记录本次 fix + feat

## 测试

- [x] CDS `pnpm tsc --noEmit` 零错误
- [x] `pnpm vitest run` 新增 7 例全绿；既有 proxy + scheduler 74 例无回归
- [ ] 真人通过预览域名验收唤醒闭环：https://modest-lovelace-h12dep-claude-prd-agent.miduo.org/ —— 这是 **CDS 代理层**改动（影响所有预览子域），验收路径为「让分支降温 → 访问其预览域名 → 应看到唤醒等待页而非诊断死页 → 几秒后落到真实应用」

> 沙箱无 docker，真实 `docker restart → 健康 → 落地真实页面` 的端到端闭环只能在 CDS 环境复测；scoping 决策已由 vitest 守住。

## 风险与已知边界

- **沙箱无 docker，未端到端验证**：scoping 有 vitest 守卫，但真实 restart 闭环需 CDS 环境首次部署后人工复测一次
- **容器被 janitor 回收后无法原地重启**：`restartServiceInPlace` 返回 false → 分支转 `error` 展示诊断页（不卡无限 spinner），用户需手动「重新部署」；未来可在等待页补一键重部署入口
- **爬虫/链接预取仍可能唤醒**：已用「仅顶层 HTML 导航请求」过滤静态资源/非 document 请求，但发送 `Accept: text/html` 的爬虫仍可能触发一次唤醒——缓解而非根除；`CDS_PREVIEW_AUTOWAKE=0` 可整体关闭
- **多服务分支**：逐个 `restartServiceInPlace`，任一失败该分支转 `error`，未做部分成功的细粒度展示
- **与调度器 tick 竞态**：靠 `branchOperationCoordinator` 租约（restart 优先级 80 > scheduler-cooling 30）抢占；唤醒成功后 `markHot` + 刷新 `lastAccessedAt`，避免下一 tick 立刻又按陈旧时间戳降温（与手动 `/restart` 同款修法）
- **行为反转**：本功能反转了「预览访问不自动启动容器」的默认设计，但严格收口在调度器降温分支；非降温（报错/崩溃/手动停/已删除）一律保留诊断页

## 后续事项

- 等待页补「一键重新部署」入口，覆盖容器已被回收的 fallback（低优先级）
- CDS 环境实测自动唤醒闭环 + janitor 回收 fallback + kill-switch 各一次（见 debt D4 完成标准）

---

&gt; [!NOTE]
&gt; **Medium Risk**
&gt; Touches preview routing and background container restarts with lease coordination; scoping limits passive wake to scheduler-cooled branches, but concurrent deploy/cool races and missing containers after janitor still need ops awareness.
&gt; 
&gt; **Overview**
&gt; Adds **preview auto-wake** so branches the scheduler cooled (`idle`, `lastStopSource=&#39;scheduler&#39;`, containers still present) are revived on **top-level HTML navigation** via in-place `docker restart`—users see the waiting page instead of a manual-redeploy dead end. `ProxyService` wires `setOnReviveCooled`, eligibility checks, and per-slug dedupe; `index.ts` implements the callback with operation leases, sync `restarting` state before the first await, `markHot` + `lastAccessedAt` on success, and **`CDS_PREVIEW_AUTOWAKE=0`** to disable.
&gt; 
&gt; Fixes preview-subdomain **diagnostic page** links (“返回 CDS 控制台”, “查看加载页预览”) by building **`https://{dashboardDomain|mainDomain}`** URLs so they no longer resolve on the preview host.
&gt; 
&gt; Adds **`proxy-auto-wake.test.ts`** (7 cases) for scoping and dedupe; changelog and debt doc **D4** note E2E and edge cases.
&gt; 
&gt; &lt;sup&gt;Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d455fa4c63cf923f70513e63d5e313a67bd9dda1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).&lt;/sup&gt;

---

> [!NOTE]
> **Medium Risk**
> Changes preview routing and background container restarts with lease coordination; scoping limits passive wake to scheduler-cooled local branches, but races with deploy/cool and janitor-recycled containers still need ops awareness.
> 
> **Overview**
> **Preview auto-wake** revives branches the scheduler cooled (`idle`, `lastStopSource='scheduler'`, containers still present) when someone opens a **preview-host** page with a real browser navigation signal—not health checks, prefetch, HEAD, assets, or non-preview routing. `ProxyService` registers `setOnReviveCooled`, eligibility checks, per-slug dedupe, and synchronous trigger so the waiting page shows progress; `index.ts` runs in-place `docker restart` under an **`auto-restart`** lease (priority 35, preempts scheduler-cooling), readiness probes, LRU eviction, `markHot` / timestamp refresh, and **`CDS_PREVIEW_AUTOWAKE=0`** to disable.
> 
> **Diagnostic UX**: failure/gone pages use **`dashboardBaseUrl()`** so console links hit `dashboardDomain` / `mainDomain` instead of the preview subdomain. The waiting-page poll **reloads** when branch-level wait hits a terminal non-loading state so failed wakes show guidance instead of an endless spinner.
> 
> **Tests & docs**: new `proxy-auto-wake.test.ts` guards scoping; changelog and debt **D4** capture E2E gaps and boundaries.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9ee460dbc7a479d302ac9459e37c4a52a378f6a4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by inernoro** [comment]
已处理本轮全部自动审查发现（均已 push 到本分支）：

| 来源 | 严重度 | 问题 | 修复 |
|---|---|---|---|
| Codex | P1 | `triggerCooledWake` 传了 v3 预览 slug 而非 `branch.id`，非 legacy 分支 `getBranch()` 查不到 → 静默不唤醒 | 改传 `branch.id`（与同函数 `resolveUpstream`/`scheduler.touch` 一致）+ 新增 canonical-id 回归用例 — `a530eb0` |
| Cursor Bugbot | High | 唤醒 `catch` 无条件把 `restarting→idle` 并标 failed，未区分 `BranchOperationSupersededError`，会覆盖抢占方的在途状态 | 对齐手动 `/restart`：superseded→`cancelled` 且不碰状态；真实失败→置 `error`（而非 idle，避免无限重试唤醒）— `8db26cc` |
| Cursor Bugbot | Medium | 未跳过远端执行器分支，本机 `restartServiceInPlace` 注定失败，把 idle 翻成 error 反而更糟 | 唤醒前检查 `executorId` 指向非 embedded 执行器则直接跳过，保留诊断页 — `282bda1` |

验证：`pnpm tsc --noEmit` 零错误；`proxy` + `proxy-auto-wake` 共 50 例全绿（含新增 canonical-id 回归）。真实 docker restart 闭环仍需 CDS 环境复测（见 `doc/debt.cds-nginx-loading-pages.md` D4）。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cds/src/index.ts [hunk 1]
```diff
@@ -1236,12 +1236,147 @@ schedulerService.setCoolFn(async (slug: string) => {
...
+
+    // Remote-executor branches run their containers off-master; the local
+    // `restartServiceInPlace` (docker on THIS host) can't reach them — manual
+    // /restart returns 409 for exactly this. Skip auto-wake (no state flip) so
+    // the branch keeps its idle diagnostic page instead of being flipped to
+    // restarting then error by a doomed local docker restart.
+    const remoteExecutor = branch.executorId && registry
+      ? registry.getAll().find((n) => n.id === branch.executorId && n.role !== 'embedded')
+      : null;
+    if (remoteExecutor) return;
```
[the remark this answers is not in the document]
Comment (inernoro) [reply cds/src/index.ts]: 已修复（`5cb27ea`）。确认了归属语义：解析完成的本地/embedded 部署会把 `executorId` 清成 `undefined`（`branches.ts`: `stillRemote?.role==='embedded' || !stillRemote → undefined`），所以 `executorId` 只要为真就一定是远端执行器——无论在线还是临时离线（coordinator 重启 / 漏 heartbeat）。两种情况本机 `restartServiceInPlace` 都够不到，统一改为 `if (branch.executorId) return` 直接跳过、不碰状态，保留诊断页，待执行器回归后由后续访问重试。守卫提到 `proxy.shouldAutoWakeCooled`（可单测，远端分支根本不触发回调），index.ts 回调保留同款二次校验；新增执行器归属分支不唤醒的用例。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cds/src/services/proxy.ts [hunk 2]
```diff
@@ -545,6 +560,24 @@ export class ProxyService {
...
+      // revive flips status synchronously so serveBranchStatusResponse below
+      // renders the live waiting page instead of the cooled dead-end.
+      // Pass the resolved entry id (branch.id) — NOT branchSlug — because for
+      // non-legacy/v3 projects the entry lives under a canonical id like
+      // `${projectSlug}-${slug}` while branchSlug is the bare preview label.
+      // The revive callback does stateService.getBranch(id), so the wrong key
+      // would silently no-op (same reason resolveUpstream/scheduler.touch below
+      // use branch.id).
+      if (this.isHtmlNavigationRequest(req) && this.shouldAutoWakeCooled(branch)) {
+        this.triggerCooledWake(branch.id);
```
[the remark this answers is not in the document]
Comment (inernoro) [reply cds/src/services/proxy.ts]: 已修复（`129d8dd`）。唤醒前新增 `req.method === 'GET'` 判定，`HEAD /` 健康探针不再触发唤醒；未改 `isHtmlNavigationRequest` 自身语义（HTML 页面渲染仍可 HEAD）。新增 HEAD 不唤醒用例。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cds/src/index.ts [hunk 3]
```diff
@@ -1236,12 +1236,150 @@ schedulerService.setCoolFn(async (slug: string) => {
...
+    let branchOperationFinalStatus: 'completed' | 'failed' | 'cancelled' = 'completed';
+
+    // Synchronous prefix (runs before the first await): flip to a loading state
+    // and persist, so the waiting page the proxy renders right after firing
+    // this callback shows live progress instead of the cooled dead-end.
+    branch.status = 'restarting';
+    for (const svc of services) {
+      if (svc.status === 'stopped' || svc.status === 'idle') svc.status = 'starting';
+    }
+    stateService.save();
```
[the remark this answers is not in the document]
Comment (inernoro) [reply cds/src/index.ts]: 已修复（`129d8dd`）。把「翻 `restarting` + 首个 `save()`」的同步前缀整体挪进 `try` 顶部（仍在首个 `await` 之前，proxy 渲染等待页的时序不变），这样首个 save 抛异常时也会走 `finally` 释放租约，不再把 `BranchOperationCoordinator` 永久挂在 active op 上。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cds/src/services/proxy.ts [hunk 4]
```diff
@@ -545,6 +560,29 @@ export class ProxyService {
...
+      // The revive callback does stateService.getBranch(id), so the wrong key
+      // would silently no-op (same reason resolveUpstream/scheduler.touch below
+      // use branch.id).
+      // Restrict the wake to real GET navigations: isHtmlNavigationRequest also
+      // accepts HEAD and a missing Accept header, so uptime monitors / link
+      // checkers doing `HEAD /` against a preview host must not restart cooled
+      // containers (they aren't a user actually opening the page).
+      const isGetNavigation = (req.method || 'GET').toUpperCase() === 'GET';
+      if (isGetNavigation && this.isHtmlNavigationRequest(req) && this.shouldAutoWakeCooled(branch)) {
+        this.triggerCooledWake(branch.id);
```
[the remark this answers is not in the document]
Comment (inernoro) [reply cds/src/services/proxy.ts]: 已修复（`5cae033`）。唤醒前新增 `isPreviewHost = extractPreviewBranch(req.headers.host) !== null` 判定——与等待页轮询 `/_cds/waiting-status` 用的是同一套 host 解析，所以「能唤醒」必然「轮询能解析、重启成功后能 reload」。非预览路由（X-Branch / cookie / 路由规则 / 默认分支）不再触发唤醒，保留诊断页。新增「非预览路由不唤醒」用例，并把测试套件整体改为经预览 host 路由。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cds/src/index.ts [hunk 5]
```diff
@@ -1236,12 +1236,154 @@ schedulerService.setCoolFn(async (slug: string) => {
...
+        lease?.assertCurrent('auto-wake before success save');
+        branch.status = 'running';
+        branch.errorMessage = undefined;
+        branch.lastStoppedAt = undefined;
+        branch.lastStopReason = undefined;
+        branch.lastStopSource = undefined;
+        // Manual /restart 同款：先刷新 lastAccessedAt 再 markHot，否则被空闲 TTL
+        // 降温过的陈旧时间戳会让下一个调度 tick 立刻又把它降温。
+        branch.lastAccessedAt = new Date().toISOString();
+        schedulerService.markHot(slug);
```
[the remark this answers is not in the document]
Comment (inernoro) [reply cds/src/index.ts:1394]: 已修复（`5cae033`）。在重启前调用 `schedulerService.evictLruIfOverCapacity(slug)`（`scheduler.wake()` 用的同一个公共方法、同样的顺序：先腾位再拉起），超过 `maxHotBranches` 时驱逐 LRU 分支，连开多个降温预览 URL 不再冲破热容器预算。`slug` 作为 excludeSlug 排除自身；调度器未启用或未配 `maxHotBranches` 时为 no-op。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cds/src/index.ts [hunk 6]
```diff
@@ -1236,12 +1236,162 @@ schedulerService.setCoolFn(async (slug: string) => {
...
+          actor: 'scheduler',
+          trigger: 'scheduler',
+          operation: 'branch-auto-wake',
+          source: 'proxy.preview-auto-wake',
+          reason: '预览访问自动唤醒（docker restart，未重建代码）',
+        });
+        lease?.assertCurrent(`auto-wake after ${svc.profileId}`);
+        if (ok) {
+          svc.status = 'running';
+          svc.errorMessage = undefined;
```
[the remark this answers is not in the document]
Comment (inernoro) [reply cds/src/index.ts]: 已修复（`5fa87bd`）。对齐 deploy 的就绪门：`restartServiceInPlace` 成功后不再立刻标 running，而是按 profile 先可选 `waitForStartupSignal`，再 `waitForReadiness(svc.hostPort, profile.readinessProbe)`（TCP+HTTP，noHttp 模式仅 TCP），真正就绪才标 running、超时标 error。等待期间分支保持 `restarting`，等待页持续轮询显示进度，直到上游真能服务才 reload 落地——消除「容器存活即 running → reload 到仍在绑定端口的上游」的 502/刷新循环。就绪等待后补一次 `assertCurrent`，长等待中被高优先级操作抢占时正确走 superseded 分支。

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by inernoro** [comment]
已处理本轮两个 Codex P2（提交 `7dc3a0a`）：

- **唤醒失败时等待页落到诊断页**（`cds/src/index.ts` 失败路径 + `proxy.ts` 轮询脚本）：唤醒的 restart 失败或就绪超时会把分支翻 `error`，但等待页轮询原来只在 `data.ready` 时 reload，`error/idle/stopped` 这种「非 loading 终态」不 reload，访问者卡在「重启中」页看不到诊断页。轮询新增 `if (data.loading === false) location.reload()`，终态 reload 后服务端渲染诊断页（带重部署指引）。这条对所有失败场景（含 deploy 失败）都是改进。
- **唤醒需正向浏览器导航信号**（`proxy.ts`）：`isHtmlNavigationRequest` 把缺失 `Accept` 当 HTML，缺 `Accept` + Fetch Metadata 的裸 `GET /` 探针仍会触发唤醒。新增 `hasBrowserNavSignal`：要求 `Accept` 含 `text/html`，或 `Sec-Fetch-Mode=navigate` / `Sec-Fetch-Dest=document`——真实浏览器导航必有其一，裸 GET 探针被挡。

测试：`proxy-auto-wake` 13 例（新增「裸 GET 无信号不唤醒」「仅 Sec-Fetch 元数据可唤醒」）；`pnpm tsc --noEmit` 零错误。

---
_Generated by [Claude Code](https://claude.ai/code)_
